File: app/src/main/java/com/tangem/data/network/BlockcypherApi.java
package com.tangem.data.network;

import com.tangem.data.network.model.BlockcypherBody;
import com.tangem.data.network.model.BlockcypherResponse;
import com.tangem.data.network.model.BlockcypherFee;
import com.tangem.data.network.model.BlockcypherTx;

import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface BlockcypherApi {
    @GET(Server.ApiBlockcypher.Method.MAIN)
    Call<BlockcypherFee> blockcypherMain(@Path("blockchain") String blockchain, @Path("network") String network, @Query("token") String token);

    @GET(Server.ApiBlockcypher.Method.ADDRESS)
    Call<BlockcypherResponse> blockcypherAddress(@Path("blockchain") String blockchain, @Path("network") String network, @Path("address") String address, @Query("token") String token);

    @GET(Server.ApiBlockcypher.Method.TXS)
    Call<BlockcypherTx> blockcypherTxs(@Path("blockchain") String blockchain, @Path("network") String network, @Path("txHash") String txHash, @Query("token") String token);

    @Headers("Content-Type: application/json")
    @POST(Server.ApiBlockcypher.Method.PUSH)
    Call<BlockcypherResponse> blockcypherPush(@Path("blockchain") String blockchain, @Path("network") String network, @Body BlockcypherBody blockcypherBody, @Query("token") String token);
}

File: blockchain/src/main/java/com/tangem/blockchain/bitcoin/network/api/BlockcypherApi.kt
package com.tangem.blockchain.bitcoin.network.api

import com.squareup.moshi.JsonClass
import com.tangem.blockchain.bitcoin.network.response.BlockcypherFee
import com.tangem.blockchain.bitcoin.network.response.BlockcypherResponse
import com.tangem.blockchain.bitcoin.network.response.BlockcypherTx
import retrofit2.http.*

interface BlockcypherApi {
    @GET("v1/{blockchain}/{network}")
    suspend fun getFee(
            @Path("blockchain") blockchain: String,
            @Path("network") network: String
    ): BlockcypherFee

    @GET("v1/{blockchain}/{network}/addrs/{address}?unspentOnly=true&includeScript=true")
    suspend fun getAddressData(
            @Path("blockchain") blockchain: String,
            @Path("network") network: String,
            @Path("address") address: String
    ): BlockcypherResponse

    @GET("v1/{blockchain}/{network}/txs/{txHash}?includeHex=true")
    suspend fun getTransactions(
            @Path("blockchain") blockchain: String,
            @Path("network") network: String,
            @Path("txHash") txHash: String
    ): BlockcypherTx

    @Headers("Content-Type: application/json")
    @POST("v1/{blockchain}/{network}/txs/push")
    suspend fun sendTransaction(
            @Path("blockchain") blockchain: String,
            @Path("network") network: String,
            @Body blockcypherBody: BlockcypherBody,
            @Query("token") token: String
    ): BlockcypherTx
}

@JsonClass(generateAdapter = true)
data class BlockcypherBody(val tx: String)